### PR TITLE
Spec compliant merge shares

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This is the first draft for spec compliant merging of shares.

TODO:
- [X] testing all edge cases
- [X] integration test
- [X] some fuzzing

Closes: #234

### CODE REVIEW

File: types/shares_test.go [hunk 1]
```diff
@@ -74,8 +75,8 @@ func TestMakeShares(t *testing.T) {
 			NamespacedShares{
 				NamespacedShare{
 					Share: append(
-						append(reservedTxNamespaceID, byte(0)),
-						zeroPadIfNecessary(smolTxLenDelimited, TxShareSize)...,
+						reservedTxNamespaceID,
+						zeroPadIfNecessary(smolTxLenDelimited, MsgShareSize)...,
```
Comment (evan-forbes) [inline types/shares_test.go:79]: I was confused about the `*` byte mentioned in the spec. I didn't know that it always had to be there. Please, ignore, I'll change this later.

File: types/block.go [hunk 2]
```diff
@@ -1637,7 +1637,8 @@ func (data *EvidenceData) splitIntoShares() NamespacedShares {
 		if err != nil {
 			panic(fmt.Sprintf("evidence included in evidence pool that can not be encoded %#v, err: %v", ev, err))
 		}
-		rawDatas = append(rawDatas, rawData)
+		delimetedRawData := append(append(make([]byte, 0, len(rawData)+1), byte(0)), rawData...)
```
Comment (liamsi) [inline types/block.go:1640]: Can we introduce a constant or at least a local variable for the `1` used above. Mostly to document what it stands for.

File: types/shares.go [hunk 3]
```diff
@@ -83,12 +88,10 @@ func splitContiguous(nid namespace.ID, rawDatas [][]byte) []NamespacedShare {
...
 		var rawData []byte
-		startIndex := 0
-		rawData, outerIndex, innerIndex, startIndex = getNextChunk(rawDatas, outerIndex, innerIndex, TxShareSize)
-		rawShare := append(append(append(
-			make([]byte, 0, len(nid)+1+len(rawData)),
+		rawData, outerIndex, innerIndex, _ = getNextChunk(rawDatas, outerIndex, innerIndex, MsgShareSize)
+		rawShare := append(append(
+			make([]byte, 0, MsgShareSize),
 			nid...),
-			byte(startIndex)),
```
Comment (liamsi) [inline types/shares.go:91]: Oh, was there a bug in the split logic?

File: types/shares.go [hunk 4]
```diff
@@ -184,3 +185,328 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
 	}
 	return share
 }
+
+// DataFromSquare extracts block data from an extended data square.
+func DataFromSquare(eds *rsmt2d.ExtendedDataSquare) (Data, error) {
```
Comment (liamsi) [inline types/shares.go:190]: 🤩

File: types/shares_test.go (hunk 1 above)
Comment (liamsi) [reply types/shares_test.go:79]: That is why I insisted on pseudo-code (or you know, an actual implementation) instead of that diagram in the spec.

**Review by liamsi — COMMENTED** [review]
Really excited about this!

File: types/shares.go (hunk 3 above)
Comment (evan-forbes) [reply types/shares.go:91]: Kinda sorta. I still have to iron out the details, but the length delimiter for contiguous Txs was being added twice. Fixing it here caused other problems elsewhere (all the modifications to the tests). We just have to clarify when and where to add the `SHARE_RESERVED_BYTE` and I think were good to go.

File: types/shares_test.go [hunk 5]
```diff
@@ -194,14 +196,188 @@ func Test_appendToSharesOverwrite(t *testing.T) {
...
+	}
+}
+
+func generateRandomlySizedContiguousShares(count, max int) Txs {
+	txs := make(Txs, count)
+	for i := 0; i < count; i++ {
+		size := rand.Intn(max)
+		// TODO: find out why
+		// txs smaller than two bytes that get mixed in with other randomly
+		// sized txs *sometime* cause processContiguousShares to end early
```
Comment (evan-forbes) [inline types/shares_test.go:267]: :grimacing: I'm hoping we can find out why during fuzzing. I'm guessing this should at least be documented somewhere? Would there ever be a scenario where transactions of less than two bytes are valid?

File: types/shares_test.go [hunk 6]
```diff
@@ -194,14 +196,188 @@ func Test_appendToSharesOverwrite(t *testing.T) {
...
+	}
+}
+
+func generateRandomlySizedContiguousShares(count, max int) Txs {
+	txs := make(Txs, count)
+	for i := 0; i < count; i++ {
+		size := rand.Intn(max)
+		// TODO: find out why
+		// txs smaller than 5 bytes that get mixed in with other randomly
+		// sized txs *sometimes* cause processContiguousShares to end early
```
Comment (evan-forbes) [inline types/shares_test.go:267]: :grimacing: I'm hoping we can find out why during fuzzing. I'm guessing this should at least be documented somewhere? Would there ever be a scenario where valid transactions, intermediate state roots, or evidence are less than 5 bytes?

**Comment by evan-forbes** [comment]
While most of the testing is finished, there still needs to be an integration test. I'm going to go ahead and mark this ready for review to start getting feedback in the mean time.

File: types/shares.go [hunk 7]
```diff
@@ -111,7 +112,7 @@ func split(rawData []byte, nid namespace.ID) []NamespacedShare {
 	for len(rawData) > 0 {
 		shareSizeOrLen := min(MsgShareSize, len(rawData))
 		rawShare := append(append(
-			make([]byte, 0, len(nid)+1+len(rawData[:shareSizeOrLen])),
+			make([]byte, 0, len(nid)+len(rawData[:shareSizeOrLen])),
```
Comment (evan-forbes) [inline types/shares.go:115]: this `+1` was mistakenly added by myself during #235

File: types/shares_test.go (hunk 6 above)
Comment (liamsi) [reply types/shares_test.go:267]: > Would there ever be a scenario where valid transactions, intermediate state roots, or evidence are less than 5 bytes?

I do not think so. cc @adlerjohn

File: types/shares.go [hunk 8]
```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+
+	ISRs := make([]tmbytes.HexBytes, len(rawISRs))
+	for i := 0; i < len(ISRs); i++ {
+		ISRs[i] = rawISRs[i]
+	}
+
+	return IntermediateStateRoots{RawRootsList: ISRs}, nil
+}
+
+// parseMsgs collects all evidence from the shares provided
```
Comment (liamsi) [inline types/shares.go:294]:
```suggestion
// parseEvd collects all evidence from the shares provided.
```

File: types/shares.go [hunk 9]
```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		return MessagesEmpty, err
+	}
+
+	return Messages{
+		MessagesList: msgList,
+	}, nil
+}
+
+func processContiguousShares(shares [][]byte) (txs [][]byte, err error) {
+	share := shares[0][NamespaceSize+ShareReservedBytes:]
```
Comment (liamsi) [inline types/shares.go:336]: Is somewhere guaranteed that len > 0?

File: types/shares.go [hunk 10]
```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		return EvidenceData{}, err
+	}
+
+	evdList := make(EvidenceList, len(rawEvd))
+
+	// parse into protobuf bytes
+	for i := 0; i < len(rawEvd); i++ {
+		// unmarshal the evidence
+		var protoEvd *tmproto.Evidence
+		err := proto.Unmarshal(rawEvd[i], protoEvd)
```
Comment (liamsi) [inline types/shares.go:307]: Maybe add a one-sentence comment on top of this explaining that the len-delimiter is already processed at this stage (in processContiguousShares)?

File: types/shares.go [hunk 11]
```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		}
+
+		txs = append(txs, newTxs...)
+
+		// if there is no next share
+		if len(shares) <= i+1 {
+			break
+		}
+
+		nextShare := shares[i+1][NamespaceSize+ShareReservedBytes:]
```
Comment (liamsi) [inline types/shares.go:356]: This is somewhat confusing to read. 

First, `i < len(shares)` because of the for-loop. This means from `len(shares) <= i+1` and `i < len(shares)` always follows `len(shares) == i+1`. So maybe change that break condition to `if len(shares) == i+1 ` (the case `len(shares) < i+1 && i < len(shares)` is impossible).

Then, add a comment on top  of this line: `// At this point we could asset: i+1 < len(shares).` Optionally, you could try to structure the code in a way that is immediately understandable without any code comments (with the clever use of variables).

File: types/shares.go [hunk 12]
```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		// if there is no current share, process the next share
+		if len(share) == 0 {
+			share, txLen, err = parseDelimiter(nextShare)
+			if err != nil {
+				break
+			}
+			continue
+		}
+
+		// create the next share by merging the next share with the extra
```
Comment (liamsi) [inline types/shares.go:367]: with the extra ...?

File: types/shares_test.go (hunk 6 above)
Comment (liamsi) [reply types/shares_test.go:267]: IMO, we can assume that Txs are at least some realistic number of bytes (probably at least ~50 bytes, but not sure). But we should still try to understand what goes wrong here (not blocking this PR though).

**Review by liamsi — COMMENTED** [review]
This looks amazing!

**Comment by liamsi** [comment]
@adlerjohn @evan-forbes now that this is (almost) ready, what do you two think about https://github.com/lazyledger/lazyledger-specs/issues/152 ?

File: types/shares_test.go (hunk 6 above)
Comment (adlerjohn) [reply types/shares_test.go:267]: > Would there ever be a scenario where valid transactions, intermediate state roots, or evidence are less than 5 bytes?

Individually, or in total? Individual ones will never be less than 5 bytes. But there could be zero txs, ISRs, and evidence (so the total bytes would be 0).

File: types/shares.go (hunk 9 above)
Comment (evan-forbes) [reply types/shares.go:344]: good catch, it is not! https://github.com/lazyledger/lazyledger-core/pull/261/commits/014a00e0f44e3c7e8c2d9ed8073e1c866c0831e5

File: types/shares.go (hunk 10 above)
Comment (evan-forbes) [reply types/shares.go:309]: I added docs to `processContiguousShare` as well https://github.com/lazyledger/lazyledger-core/pull/261/commits/014a00e0f44e3c7e8c2d9ed8073e1c866c0831e5

File: types/shares.go (hunk 11 above)
Comment (evan-forbes) [reply types/shares.go]: It's definitely difficult to read. I've made some of the changes suggested, but I still think it can be improved more. I struggled with making this more readable the first try too. I'll continue to chew on this while I write the integration test and fix the occasional super small tx size bug. https://github.com/lazyledger/lazyledger-core/pull/261/commits/014a00e0f44e3c7e8c2d9ed8073e1c866c0831e5

File: types/shares.go (hunk 12 above)
Comment (evan-forbes) [reply types/shares.go]: the remaining portion of the share. changed in https://github.com/lazyledger/lazyledger-core/pull/261/commits/014a00e0f44e3c7e8c2d9ed8073e1c866c0831e5

File: types/shares_test.go (hunk 6 above)
Comment (evan-forbes) [reply types/shares_test.go:267]: > Individually, or in total?

individually. 

That makes sense. I'll continue to investigate

File: types/shares.go [hunk 13]
```diff
@@ -86,7 +89,7 @@ func splitContiguous(nid namespace.ID, rawDatas [][]byte) []NamespacedShare {
 		startIndex := 0
 		rawData, outerIndex, innerIndex, startIndex = getNextChunk(rawDatas, outerIndex, innerIndex, TxShareSize)
 		rawShare := append(append(append(
-			make([]byte, 0, len(nid)+1+len(rawData)),
+			make([]byte, 0, len(nid)+ShareReservedBytes+len(rawData)),
```
Comment (adlerjohn) [inline types/shares.go:92]: Probably easier to use the `ShareSize` constant directly.

```suggestion
			make([]byte, 0, ShareSize),
```

File: types/shares.go (hunk 7 above)
Comment (adlerjohn) [inline types/shares.go:115]:
```suggestion
			make([]byte, 0, ShareSize),
```

File: types/shares.go [hunk 14]
```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+			case bytes.Equal(EvidenceNamespaceID, nid):
+				evdShares = append(evdShares, share)
+
+			case bytes.Equal(TailPaddingNamespaceID, nid):
+				continue
+
+			// every other namespaceID should be a message
+			default:
+				msgShares = append(msgShares, share)
+			}
```
Comment (adlerjohn) [inline types/shares.go:229]: This logic doesn't seem to conform to specs. Potential namespace IDs: https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/consensus.md#reserved-namespace-ids. Namespaces up to `NAMESPACE_ID_MAX_RESERVED = 255` are reserved, not messages: https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/consensus.md#constants.

There should also be a check _somewhere_ that a reserved namespace ID that's up to 255 but not specifically reserved is invalid.

File: types/shares.go [hunk 15]
```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		}
+	}
+
+	// pass the raw share data to their respective parsers
+	txs, err := parseTxs(txsShares)
+	if err != nil {
+		return Data{}, err
+	}
+
+	isrs, err := parseIsrs(isrShares)
```
Comment (adlerjohn) [inline types/shares.go:239]: Inconsistent capitalization on "Isr" "ISR"

File: types/shares.go [hunk 16]
```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		MessagesList: msgList,
+	}, nil
+}
+
+// processContiguousShares takes raw shares and extracts out transactions,
+// intermediate state roots, or evidence. The returned [][]byte do have
+// namespaces or length delimiters and are ready to be unmarshalled
+func processContiguousShares(shares [][]byte) (txs [][]byte, err error) {
+	if len(shares) == 0 {
+		return nil, err
```
Comment (adlerjohn) [inline types/shares.go:342]:
```suggestion
		return nil, nil
```

File: types/shares.go [hunk 17]
```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+}
+
+func nextMsg(shares [][]byte, current, nid []byte, cursor, l uint64) ([]byte, []byte, uint64, uint64, Message, error) {
+	switch {
+	// the message uses all of the current share data and at least some of the
+	// next share
+	case l > uint64(len(current)):
+		// add the next share to the current one and try again
+		cursor++
+		current := append(current, shares[cursor][NamespaceSize:]...)
```
Comment (adlerjohn) [inline types/shares.go:449]: Nit, but there doesn't seem a need to allocate a new pointer here.

```suggestion
		current = append(current, shares[cursor][NamespaceSize:]...)
```

File: types/shares.go [hunk 18]
```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+	}
+
+	return msgs, nil
+}
+
+func nextMsg(shares [][]byte, current, nid []byte, cursor, l uint64) ([]byte, []byte, uint64, uint64, Message, error) {
+	switch {
+	// the message uses all of the current share data and at least some of the
+	// next share
+	case l > uint64(len(current)):
```
Comment (adlerjohn) [inline types/shares.go:446]: What is the reason for the magic `1`? I'm not clear on what this code is doing.

Comment (evan-forbes) [reply types/shares.go]: yikes, that is a terrible var name :sweat_smile: what about `msgLen`? https://github.com/lazyledger/lazyledger-core/pull/261/commits/8794411f54caf3ec87f6c3378d990b736ba6ffab

File: types/shares.go (hunk 14 above)
Comment (evan-forbes) [reply types/shares.go:233]: Thanks for pointing this out! added issue {#67](https://github.com/lazyledger/lazyledger-app/issues/67) to track in the lazyledger-app to add a check in `MsgWirePayForMessage`'s `ValidateBasic` method which get's called with every call to `CheckTx`. 

Added a `case` for this along with a namespace var in `consts.go` to represent the max namespaceID https://github.com/lazyledger/lazyledger-core/blob/d118716769036925491d0412772693a55546ca2c/types/shares.go#L227

Should I also add the const `NamespaceIDMaxReserved` and use that to construct `MaxReservedNamespaceID`?

File: types/shares.go (hunk 15 above)
Comment (evan-forbes) [reply types/shares.go]: https://github.com/lazyledger/lazyledger-core/pull/261/commits/39ce26a00b48a1d706bf42b93a0a43d07d8193ba

File: types/shares.go (hunk 14 above)
Comment (adlerjohn) [reply types/shares.go:233]: > Should I also add the const `NamespaceIDMaxReserved` and use that to construct `MaxReservedNamespaceID`?

Nah, the setup you have now should be sufficient I think.

Comment (adlerjohn) [reply types/shares.go:233]: One additional check that needs to be _somewhere_ is that padding shares between a transaction and a message must have namespace ID `TAIL_TRANSACTION_PADDING_NAMESPACE_ID` (all the way at the bottom of https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/data_structures.md#arranging-available-data-into-shares, under the giant figures)

> The non-interactive default rules may introduce empty shares that do not belong to any message (in the example above, the top-right share is empty). These are zeroes with namespace ID equal to the either `TAIL_TRANSACTION_PADDING_NAMESPACE_ID` if between a request with a reserved namespace ID and a message, or the namespace ID of the previous message if succeeded by a message. See the rationale doc for more info.

**Comment by evan-forbes** [comment]
So, I'm still debugging the last bug. The tests pass most of the time, but they occasionally fail when randomly sized txs, evidence, ISRs, and messages are used. Will continue to debug after a break.

**Comment by evan-forbes** [comment]
The CI linter is failing for some weird reason. I can't replicate this locally, and it's also not true... `ComputeShares` only returns a single value.
```
Can't run linter goanalysis_metalinter: bodyclose: failed prerequisites: [buildssa@github.com/lazyledger/lazyledger-core/types [github.com/lazyledger/lazyledger-core/types.test]: analysis skipped: errors in package: [/home/runner/work/lazyledger-core/lazyledger-core/types/shares_test.go:242:14: 2-valued data.ComputeShares() (value of type (NamespacedShares, int)) where single value is expected]]"
```

**Comment by liamsi** [comment]
> ComputeShares only returns a single value.

Not sure what is going on with the linter but did you merge in master (because #249 / #262).

**Comment by evan-forbes** [comment]
There were a few minor sneaky bugs. The first being that the implementation wasn't planning for the length delimiter occasionally getting split up between shares. The second was that the start index (`SHARE_RESERVED_BYTE`) was occasionally wrong. Lastly, the evidence needed to be marshalled using `*tmproto.Evidence` instead of `*tmproto.DuplicateVoteEvidence`.

On the third refactor, it has risen.

I also added fuzzing.

File: types/shares.go [hunk 19]
```diff
@@ -184,3 +192,327 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
 	}
 	return share
 }
+
+// /////////////////////////////
+// Merging
+// ////////////////////////////
```
Comment (liamsi) [inline types/shares.go:198]: IMO, we should split out the encoding, or rather the merge-split logic into separate files, e.g. share_split.go, shares_merge.go. Or, shares_encoding.go if we wanted both directions in one file.

File: types/shares.go [hunk 20]
```diff
@@ -184,3 +192,327 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+			return err
+		}
+		if txLen == 0 {
+			return nil
+		}
+		ss.txLen = txLen
+	}
+	safeLen := len(share) - binary.MaxVarintLen64
+	if safeLen < 0 {
+		safeLen = 0
```
Comment (liamsi) [inline types/shares.go:399]: What happens in the case `safeLen == 0`. It's non-obvious IMO.

File: types/shares.go [hunk 21]
```diff
@@ -184,3 +192,327 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+		ss.txs = append(ss.txs, share[:ss.txLen])
+		share = share[ss.txLen:]
+		return ss.peel(share, true)
+	}
+	return errors.New("failure to parse block data: transaction length exceeded data length")
+}
+
+// parseMessages iterates through raw shares and separates the contiguous chunks
+// of data. we use this for transactions, evidence, and intermediate state roots
+func parseMsgShares(shares [][]byte) ([]Message, error) {
```
Comment (liamsi) [inline types/shares.go:421]: comment is out of date:
```suggestion
// parseMsgShares iterates through raw shares and separates the contiguous chunks
// of data. It is only used for Messages, i.e. shares with a non-reserved namespace.
func parseMsgShares(shares [][]byte) ([]Message, error) {
```

**Review by liamsi — COMMENTED** [review]
This is so dope! Left some minor comments and suggestions.

File: types/shares.go [hunk 22]
```diff
@@ -184,3 +192,327 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
...
+	// define a slice for the raw share data of each type
+	var (
+		// transactions
+		txsShares [][]byte
+		// intermediate state roots
+		isrShares [][]byte
+		// evidence
+		evdShares [][]byte
+		// messages
+		msgShares [][]byte
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply types/shares.go]: I agree. https://github.com/lazyledger/lazyledger-core/pull/261/commits/76d7a4b4cbd24e717069a2aea6ca07804fcc1f82 and https://github.com/lazyledger/lazyledger-core/pull/261/commits/a88db3b3694ade4212d81369a3785f78602a85b1

File: types/shares.go (hunk 19 above)
Comment (evan-forbes) [reply types/shares.go]: good idea! https://github.com/lazyledger/lazyledger-core/pull/261/commits/1cb40309ed9a65b5c18b8fc8c38491c0a0fc84d8

File: types/shares.go (hunk 20 above)
Comment (evan-forbes) [reply types/shares.go]: that does need a comment, thanks! https://github.com/lazyledger/lazyledger-core/pull/261/commits/043812d991643bfec000590a9371b95dd7ff389e

File: types/consts.go [hunk 23]
```diff
@@ -36,6 +36,7 @@ var (
 	TxNamespaceID                     = namespace.ID{0, 0, 0, 0, 0, 0, 0, 1}
 	IntermediateStateRootsNamespaceID = namespace.ID{0, 0, 0, 0, 0, 0, 0, 2}
 	EvidenceNamespaceID               = namespace.ID{0, 0, 0, 0, 0, 0, 0, 3}
+	MaxReservedNamespace              = namespace.ID{0, 0, 0, 0, 0, 0, 0, 255}
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply types/consts.go]: went ahead and added links plus basic docs while we're talking about it https://github.com/lazyledger/lazyledger-core/pull/261/commits/887aa0885c1a53d44b36a6240c2c9e90280e9df8

**Review by liamsi — APPROVED** [review]
❤️ This is awesome! 🚀
